entsoe-balances-adjustment/src/main/java/com/powsybl/entsoe/balances_adjustment/util/CgmesVoltageLevelsArea.java

```diff
@@ -0,0 +1,130 @@
+ */
+class CgmesVoltageLevelsArea implements NetworkArea {
+
+    private final List<String> voltageLevelIds = new ArrayList<>();
+
+    private final List<DanglingLine> danglingLineBordersCache;
+    private final List<Branch<?>> branchBordersCache;
+
+    private final Set<Bus> busesCache;
+
+    CgmesVoltageLevelsArea(Network network, CgmesControlArea area, List<String> excludedXnodes, List<String> voltageLevelIds) {
+        this.voltageLevelIds.addAll(voltageLevelIds);
+
+        danglingLineBordersCache = createDanglingLinesCache(network, area, excludedXnodes);
+        branchBordersCache = createBranchesCache(network, area, excludedXnodes);
+
+        busesCache = network.getBusView().getBusStream()
+                .filter(bus -> voltageLevelIds.contains(bus.getVoltageLevel().getId()))
+                .filter(Bus::isInMainSynchronousComponent)
+                .collect(Collectors.toSet());
+    }
+
+    private List<DanglingLine> createDanglingLinesCache(Network network, CgmesControlArea area, List<String> excludedXnodes) {
+        return network.getDanglingLineStream()
+                .filter(this::isAreaBorder)
+                .filter(dl -> dl.getTerminal().getBusView().getBus() != null && dl.getTerminal().getBusView().getBus().isInMainSynchronousComponent())
+                .filter(dl -> dl.getExtension(CgmesDanglingLineBoundaryNode.class) == null || !dl.getExtension(CgmesDanglingLineBoundaryNode.class).isHvdc())
+                .filter(dl -> {
+                    if (area != null && (!area.getTerminals().isEmpty() || !area.getBoundaries().isEmpty())) {
+                        return area.getTerminals().stream().anyMatch(t -> t.getConnectable().getId().equals(dl.getId())) || area.getBoundaries().stream().anyMatch(bd -> bd.getConnectable().getId().equals(dl.getId()));
+                    }
+                    return true;
+                })
+                .filter(dl -> {
+                    if (excludedXnodes != null) {
+                        return excludedXnodes.stream().noneMatch(xnodeCode -> dl.getUcteXnodeCode().equals(xnodeCode));
+                    }
+                    return true;
+                })
+                .collect(Collectors.toList());
+    }
+
+    private List<Branch<?>> createBranchesCache(Network network, CgmesControlArea area, List<String> excludedXnodes) {
+        return network.getLineStream()
+                .filter(this::isAreaBorder)
+                .filter(b -> b.getTerminal1().getBusView().getBus() != null && b.getTerminal1().getBusView().getBus().isInMainSynchronousComponent()
+                        && b.getTerminal2().getBusView().getBus() != null && b.getTerminal2().getBusView().getBus().isInMainSynchronousComponent())
+                .filter(b -> b.getExtension(CgmesLineBoundaryNode.class) == null || !b.getExtension(CgmesLineBoundaryNode.class).isHvdc())
+                .filter(b -> !b.hasProperty("isHvdc")) // necessary as extensions of merged lines are not well handled
```

Comment:
Can you add a `FIXME` here. And I will add some comments to explicit here our filters.